File: producer/producer.py
import socket
import json
import time
import psutil
import os

BROKER_HOST = os.getenv("BROKER_HOST", "broker1")
PORT = 5000
PRODUCER_ID = os.getenv("HOSTNAME", "producer")

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

while True:
    try:
        s.connect((BROKER_HOST, PORT))
        print("✅ Productor conectado al broker", flush=True)
        break
    except:
        print("⏳ Esperando broker...", flush=True)
        time.sleep(2)

def send(topic, value):
    msg = {
        "type": "publish",
        "topic": topic,
        "key": PRODUCER_ID,
        "value": value
    }
    s.sendall((json.dumps(msg) + "\n").encode())
    print(f"📤 Enviado {topic}: {value}", flush=True)

while True:
    send("metrics/docker/cpu", f"{psutil.cpu_percent()}%")
    send("metrics/docker/memory", f"{psutil.virtual_memory().percent}%")
    time.sleep(5)

File: producer/producer.c
#include "producer.h"
#include "common/protocol.h"

#include <stdio.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <stdlib.h>

int main() {
    setbuf(stdout, NULL);

    int sock;
    struct sockaddr_in addr;
    handshake_t h;
    message_t msg;

    sock = socket(AF_INET, SOCK_STREAM, 0);

    addr.sin_family = AF_INET;
    addr.sin_port = htons(BROKER_PORT);
    inet_pton(AF_INET, "broker", &addr.sin_addr);

    connect(sock, (struct sockaddr*)&addr, sizeof(addr));

    h.role = ROLE_PRODUCER;
    send(sock, &h, sizeof(h), 0);

    while (1) {
        snprintf(msg.topic, sizeof(msg.topic), "metrics/docker/cpu");
        snprintf(msg.payload, sizeof(msg.payload), "%d%%", rand() % 100);
        send(sock, &msg, sizeof(msg), 0);
        sleep(2);
    }
}
